### PR TITLE
130/cancellation

### LINKED ISSUE #130: Implement cancelation of orders
> Part of https://github.com/gnosis/gp-ui/issues/55

Add support for cancelation of orders

Hard/Soft/both cancelation?

Hard => SC
Soft => API

**Workflow:**
- close/trash/💣/ icon on pending meta-tx
- Popup with confirmation: Soft / Hard cancelation
       title: Cancel order
       content: make proposal explainning that one is free, one u pay, one u trust the operator, one u don't. Non of them gives u arantees of cancelation.
- Soft: DELETE "/orders/<id>" call to the API
- Hard: Send tx to SC.

`EDIT:` Probably is best to start with soft cancelations. TBD how we authenticate the user. There must be some kind of authentication (include a signature in the deletion headers or sth)

**Discussion:**
https://gnosisinc.slack.com/archives/CPZA1AGKY/p1613139211144100

**Specs:**
It would be good to add the specs too
https://docs.google.com/document/d/1ko1oDAECz3Oy1g0MxdrBzEuUsMhKcRHr8mMH-yTCzfs/edit

**Related**:
* Effective balance and Effective Orders: https://github.com/gnosis/gp-swap-ui/issues/151
* Backend: Handle multiple orders for same user: https://github.com/gnosis/gp-v2-services#309

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Closes #130 

Soft order cancellation

- Pending orders can now be "soft" cancelled from the "account details" modal
- Added a "cancel" button to pending orders
screenshot_2021-06-02_15-22-24
- Cancelled orders are displayed on "account details" modal. They have the same pill color as expired orders, but the icon is the x-circle
- Notification for cancelled orders
screenshot_2021-06-02_15-27-51
- Confirmation for orders to be cancelled
screenshot_2021-06-02_15-12-57
- Refactored operator API HTTP methods
- Added sendOrderCancellation to operator API
- Refactored signature methods
- Added signOrderCancellation to signature methods

# Testing
1. Place an order
2. While it's still pending, click on `Pending` (your account details)
3. On the right of the pending order, click on `cancel`
- [ ]  On the open modal, verify the details match the order you want to cancel
4. Cancel order
5. On your wallet, sign the order cancellation
- [ ] Once it's signed, verify you see a order cancelled toast notification
- [ ] Verify on "account details" modal that the order has a red pill on the left side and red circle with an x on the right side
6. Click on the explorer link for the cancelled order
- [ ] Verify it has the status `cancelled`

## Not included in this PR:
- [x] Improve error handling from API responses (https://github.com/gnosis/cowswap/pull/737)
- [x] Improve styles for cancellable orders on "account details" modal
- [x] Improve styles for cancel order modal
- [x] Improve text for cancel order modal

### CODE REVIEW

**Review by anxolin — COMMENTED** [review]
Awesome! 

I will review tomorrow. It works great! 

Minor early feedback

Maybe that icon signals that something went wrong. I see why you added it, because is a cancelation, but the operation itself succeded. Otherwise i think something went wrong.
image


This looks like a button, maybe @biocom has some ideas, we can address in other PRs these things: 
image

The confirmation is GREAT, i like the info there. I would just hide this information under a toggle (+more). Is too verbose. 
image

Also, we might want to handle better the CANCELING state:
image

And totally only good to have (not that imp), but maybe we should have a sound for successful cancelation or use one of the one we have. 

Explorer and workflow worked like a charm! Amazing

**Comment by alfetopito** [comment]
> Maybe that icon signals that something went wrong. I see why you added it, because is a cancelation, but the operation itself succeded. Otherwise i think something went wrong.

I did that at first, but it felt odd to have the "success" check mark for a cancelled order.
I'll change it again

> This looks like a button, maybe @biocom has some ideas, we can address in other PRs these things:

True, does look like a button... :shrug: 

> The confirmation is GREAT, i like the info there. I would just hide this information under a toggle (+more). Is too verbose.

I can do that

> Also, we might want to handle better the CANCELING state:

You mean the UI when the `cancel` button shoves the loader aside? Yep, agree. I'll leave this up to professionals, though.

> And totally only good to have (not that imp), but maybe we should have a sound for successful cancelation or use one of the one we have.

Well, ok I guess. Use what, the same as expired orders?

> Explorer and workflow worked like a charm! Amazing

\o/

**Comment by alfetopito** [comment]
Update: change the pill color of cancelled orders as well as the icon for cancelled orders
screenshot_2021-06-04_09-26-10

The pill is now grey: red indicated error/failure, used for expired orders. Cancellation is not an error nor a failure. It's instead a user input, so I think grey is a neutral action color.

The icon is now green: same color as successful operations like mined txs. When the operation succeeds, marking it red would give the wrong impression.

Of course, totally up for discussion, I'm making stuff up here =P

File: src/custom/components/AccountDetails/Transaction.tsx [hunk 1]
```diff
@@ -63,22 +83,108 @@ const TransactionState = styled(OldTransactionState).attrs(
...
+  isOpen: boolean
+  orderId: string
+  summary: string | undefined
+}
+
+function CancellationModal(props: CancellationModalProps): JSX.Element | null {
+  const { orderId, isOpen, onDismiss, summary } = props
+  const shortId = shortenOrderId(orderId)
+
+  const [status, setStatus] = useState<'not started' | 'waiting for wallet' | 'error'>('not started')
```
Comment (W3stside) [inline src/custom/components/AccountDetails/Transaction.tsx:97]: make a type?

File: src/custom/components/AccountDetails/Transaction.tsx [hunk 2]
```diff
@@ -63,22 +83,108 @@ const TransactionState = styled(OldTransactionState).attrs(
...
+  }, [orderId])
+
+  const onClick = useCallback(() => {
+    setStatus('waiting for wallet')
+    setError('')
+    cancelOrder(orderId)
+      .then(onDismiss)
+      .catch(e => {
+        setError(e.message)
+        setStatus('error')
```
Comment (W3stside) [inline src/custom/components/AccountDetails/Transaction.tsx:116]: why do u need both status for error and an error? doesn't seem like u use that status apart from boolean checking if there should be an error displayed. could make error null and check like that

File: src/custom/components/AccountDetails/Transaction.tsx [hunk 3]
```diff
@@ -63,22 +83,108 @@ const TransactionState = styled(OldTransactionState).attrs(
...
+        <TransactionErrorContent onDismiss={onDismiss} message={error || 'Failed to cancel order'} />
+      ) : (
+        <>
+          <ConfirmationModalContent
+            title="Cancel order"
+            onDismiss={onDismiss}
+            topContent={() => (
+              <>
+                <p>
+                  Are you sure you want to cancel the order <em>{shortId}</em>?
```
Comment (W3stside) [inline src/custom/components/AccountDetails/Transaction.tsx:137]:
```suggestion
                  Are you sure you want to cancel order <em>{shortId}</em>?
```

File: src/custom/components/AccountDetails/Transaction.tsx [hunk 4]
```diff
@@ -63,22 +83,108 @@ const TransactionState = styled(OldTransactionState).attrs(
...
+                </p>
+                <p>{summary}</p>
+                <p>
+                  Keep in mind this is a soft cancellation{' '}
+                  <LinkStyledButton onClick={toggleShowMore}>[{showMore ? '- less' : '+ more'}]</LinkStyledButton>
+                </p>
+                {showMore && (
+                  <>
+                    <p>It will be taken into account in a best effort basis.</p>
```
Comment (W3stside) [inline src/custom/components/AccountDetails/Transaction.tsx:146]: mmm not clear what this means

File: src/custom/hooks/useCancelOrder.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+export const useCancelOrder = () => {
+  const { account, chainId, library } = useActiveWeb3React()
+  const cancelPendingOrder = useCancelPendingOrder()
+  return useCallback(
+    async (orderId: string): Promise<void> => {
+      if (!account || !chainId || !library) {
+        return
+      }
+      const signer = library.getSigner()
+      await sendOrderCancellation({ chainId, orderId, signer, account, cancelPendingOrder })
```
Comment (W3stside) [inline src/custom/hooks/useCancelOrder.ts:15]: can just return sendOrderCancellation as fn is async

**Review by W3stside — APPROVED** [review]
appriving with comments

File: src/custom/components/AccountDetails/Transaction.tsx (hunk 4 above)
Comment (alfetopito) [reply src/custom/components/AccountDetails/Transaction.tsx:146]: You mean the sentence or the logic?

For the sentence, means that you can (try to) cancel, but we don't guarantee it'll be cancelled.

Not good?

File: src/custom/components/AccountDetails/Transaction.tsx (hunk 2 above)
Comment (alfetopito) [reply src/custom/components/AccountDetails/Transaction.tsx:116]: Makes sense, will update

File: src/custom/components/AccountDetails/Transaction.tsx (hunk 4 above)
Comment (alfetopito) [reply src/custom/components/AccountDetails/Transaction.tsx]: I've removed this bit on https://github.com/gnosis/cowswap/pull/760

**Review by anxolin — COMMENTED** [review]
One small issue I had while testing this is that I saw "The order have been canceled" but when I clicked the tx link, I saw in the explorer that the order was filled. 

Maybe the message shouldn't be so assertive. It should be like `Soft cancelation requested` 

Another detail is the FAQ link. It makes you loose the context in a pretty delicate moment. 
I think it should open in a blank page:
image

File: src/custom/state/orders/helpers.tsx [hunk 6]
```diff
@@ -46,7 +50,35 @@ type MetaPopupContent = GPPopupContent<OrderTxTypes.METATXN>
...
+  }
+
+  & > p:last-chid {
+    margin-bottom: 0;
+  }
+`
+
+export function buildCancellationPopupSummary(id: string, summary: string | undefined): JSX.Element {
+  // TODO: style this!
+  // TODO: this probably shouldn't be here anyway :shrug:
```
Comment (anxolin) [inline src/custom/state/orders/helpers.tsx:72]: I think styles are mostly fine. So maybe we should remove this? or u want to make a suggestion in an issue? 

One thing that we might want to improve, is the cancel button, although I guess is also kind of fine. But the popup looks great to me.

File: src/custom/state/orders/middleware.ts [hunk 7]
```diff
@@ -172,6 +182,9 @@ export const soundMiddleware: Middleware<{}, AppState> = store => next => action
     getMoooooSuccess().play()
   } else if (isExpireOrdersAction(action)) {
     getMoooooError().play()
+  } else if (isCancelOrderAction(action)) {
+    // TODO: find a unique sound for order cancellation
+    getMoooooError().play()
```
Comment (anxolin) [inline src/custom/state/orders/middleware.ts:187]: @bh2smith :)

**Review by anxolin — APPROVED** [review]
Other than the minor details mentioned above. I think we can merge to develop already

**Comment by alfetopito** [comment]
> One small issue I had while testing this is that I saw "The order have been canceled" but when I clicked the tx link, I saw in the explorer that the order was filled.

That is changed on https://github.com/gnosis/cowswap/pull/753

> 
> Maybe the message shouldn't be so assertive. It should be like `Soft cancelation requested`
> 
> Another detail is the FAQ link. It makes you loose the context in a pretty delicate moment.
> I think it should open in a blank page:

I've changed it.
The issue with this is that, on new pages there's no scroll to anchor. It'll be at the top of the page.

**Comment by alfetopito** [comment]
Opsie, mergify went ahead and merged it :shrug: 

Work continues on https://github.com/gnosis/cowswap/pull/753

**Comment by bh2smith** [comment]
> Opsie, mergify went ahead and merged it :shrug: 

It seems this is more of an inconvenience than anything else. Have we considered removing this?

**Comment by alfetopito** [comment]
> > Opsie, mergify went ahead and merged it shrug
> 
> It seems this is more of an inconvenience than anything else. Have we considered removing this?

Nah, not a big deal on merging it.
And we can disabled by using a tag if we want.

**Pull request merged**